src/Eshopworld.WorkerProcess/LeaseAllocator.cs

```diff
@@ -34,83 +35,103 @@ public class LeaseAllocator : ILeaseAllocator
+                };
+                var createResult = await TryAcquireLease(lease, instanceId, _leaseStore.TryCreateLeaseAsync).ConfigureAwait(false);
+                if (createResult.Result || createResult.Lease==null)
+                {
+                    return createResult.Lease;
+                }
 
-                    if (updateResult.Result)
-                    {
-                        _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
-                            _options.Value.Priority,
-                            $"Lease acquired. {GenerateLeaseInfo(updateResult.Lease)}"));
+                lease = createResult.Lease;
+            }
 
-                        return updateResult.Lease;
-                    }
 
-                    _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
-                        _options.Value.Priority,
-                        $"Lease activation failed lease already acquired. Allocated Lease Id: [{lease.Id}]"));
-                }
+            lease.Priority = _options.Value.Priority;
+            lease.InstanceId = instanceId;
+            var updateTime = ServerDateTime.UtcNow;
+            lease.Interval = _slottedInterval.Calculate(updateTime, _options.Value.LeaseInterval);
+            lease.LeasedUntil = updateTime.Add(lease.Interval.Value);
+            var updateResult = await TryAcquireLease(lease, instanceId, _leaseStore.TryUpdateLeaseAsync).ConfigureAwait(false);
+            if (updateResult.Result || updateResult.Lease == null)
+            {
+                return updateResult.Lease;
+            }
+            _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
+                _options.Value.Priority,
+                $"Lease failed to be acquired on update. {GenerateLeaseInfo(updateResult.Lease)}"));
+            return null;
+           
+        }
 
-                // Read the lease for event
-                lease = await _leaseStore.ReadByLeaseTypeAsync(_options.Value.WorkerType).ConfigureAwait(false);
+        private async Task<LeaseStoreResult> TryAcquireLease(ILease lease, Guid instanceId, Func<ILease,Task<LeaseStoreResult>> leaseAcquirer)
+        {
+            var leaseResult = await leaseAcquirer(lease).ConfigureAwait(false);
 
+            if (leaseResult.Result)
+            {
                 _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
                     _options.Value.Priority,
-                    $"Lease acquisition failed lease already acquired. {GenerateLeaseInfo(lease)}"));
+                    $"Lease acquired. {GenerateLeaseInfo(leaseResult.Lease)}"));
+                return leaseResult;
             }
-            else
+
+            // Read the lease for event
+            lease = await _leaseStore.ReadByLeaseTypeAsync(_options.Value.WorkerType).ConfigureAwait(false);
```

Comment:
Why do we read it again?

Reply:
this was introduced to get the actual lease in case of an etag conflict occurs

Reply:
see above discussion